src/Control/Biegunka/Source/Git.hs

```diff
@@ -97,29 +94,43 @@ git_ :: URI -> FilePath -> Script 'Sources ()
 git_ u p = git' u p mempty
 
 updateGit :: URI -> FilePath -> Git -> IO (Maybe String)
-updateGit u p Git { _remote, _branch } =
+updateGit u p Git { _branch, _failIfAhead } =
   doesDirectoryExist p >>= \case
     True -> do
-      let rbr = _remote </> _branch
+      let rbr = "origin" </> _branch
       before <- gitHash
+      remotes <- lines `fmap` askGit p ["remote"]
+      when ("origin" `notElem` remotes) $
+        askGit' p ["remote", "add", "origin", u]
       -- Typical git: <https://stackoverflow.com/questions/11347712/git-fetch-only-for-current-branch>
-      _ <- askGit p ["fetch", _remote, _branch] -- TODO: track upstream
-      _ <- askGit p ["checkout", "-B", _branch, "--track", rbr] -- TODO: track upstream
-      _ <- askGit p ["merge", rbr, _branch] -- TODO: rebase
+      askGit' p ["fetch", "origin", _branch]
+      currentBranch <- (listToMaybe . lines) `fmap` askGit p ["rev-parse", "--abbrev-ref", "HEAD"]
+      when (currentBranch /= Just _branch) $
+        askGit' p ["checkout", "-B", _branch, "--track", rbr]
+      commitsAhead <- (not . null . lines) `fmap` askGit p ["rev-list", _branch ++ "..origin"]
+      if (commitsAhead && _failIfAhead)
+        then sourceFailure (Text.pack "local branch is ahead of remote")
+        else askGit' p ["rebase", rbr]
       after <- gitHash
       return (bool (Just (printf "‘%s’ → ‘%s’" before after)) Nothing (before == after))
     False -> do
-      _ <- askGit "/" ["clone", u, p, "-b", _branch]
+      askGit' "/" ["clone", u, p, "-b", _branch]
       after <- gitHash
       return (Just (printf "checked ‘%s’ out" after))
  where
-  gitHash = fmap (Text.unpack . Text.stripEnd) (askGit p ["rev-parse", "--short", "HEAD"])
-
-  askGit workingDirectory args = bracket
-    getCurrentDirectory
-    setCurrentDirectory $ \_ -> do
-      setCurrentDirectory workingDirectory
-      (exitcode, out, err) <- readProcessWithExitCode "git" args []
-      exitcode `onFailure`
-        \_ -> sourceFailure (Text.pack err)
-      return (Text.pack out)
+  gitHash = (askGit p ["rev-parse", "--short", "HEAD"])
+
+  askGit' :: FilePath -> [String] -> IO ()
+  askGit' f args = void $ askGit f args
+
+  askGit :: FilePath -> [String] -> IO String
+  askGit workingDirectory args = (Text.unpack . Text.stripEnd) `fmap` go
+   where
+    go = bracket
```

Comment:
I've just realized how embarrassingly bad this code is. We should only set working directory for the forked `git` process, otherwise it's unknown where it's actually run since we update many sources concurrently.

It has been broken since 73a3886b7d8a7c6574d4f9b8c0039083bc525ea1.